### ISSUE TITLE
Verify: Sequence packing attention isolation fix

### ISSUE THREAD

**Issue by Lev-Stambler** [DeepOpt-com/WrinkleFree-Monorepo#53]
## Summary

Fixed a bug where sequence packing was not properly isolating documents. The previous implementation only reset position IDs at document boundaries, but this does NOT block cross-document attention - it only affects RoPE positional encoding.

## The Problem

With position reset only (old behavior):
- Doc2 tokens could attend to Doc1 tokens
- Changing Doc1 changed Doc2 logits by up to **11.34**
- Cross-document attention sum: **0.86** (significant leakage)

## The Fix

Added `compute_document_attention_mask()` that creates a 2D block-diagonal causal mask:

```
[Doc1: A B C EOS] [Doc2: D E]

Attention mask:
      A   B   C  EOS  D   E
A  :  1   0   0   0   0   0
B  :  1   1   0   0   0   0
C  :  1   1   1   0   0   0
EOS:  1   1   1   1   0   0
D  :  0   0   0   0   1   0  ← Can't see Doc1!
E  :  0   0   0   0   1   1
```

## Verification

With the new mask:
- Doc2 logits max diff: **0.00000000** (perfect isolation)
- All 27 tests pass

## Files Changed

- `packages/data_handler/src/wf_data/data/packing.py` - Added `compute_document_attention_mask()`
- `packages/data_handler/src/wf_data/data/mixing.py` - Added `use_document_mask` parameter (default: True)
- `packages/data_handler/tests/test_attention_isolation.py` - New test file

## Action Requested

Please review and verify:
1. Is the mask computation correct?
2. Does transformers properly use 4D attention masks with SDPA/FlashAttention?
3. Should we add a config option to disable the mask for backwards compatibility?

## Reference

- [HuggingFace blog on sequence packing](https://huggingface.co/blog/sirluk/llm-sequence-packing)

**Comment by Lev-Stambler** [comment]
## Fixed: Mask Dtype Issue

Updated `compute_document_attention_mask()` to return float mask directly:
- `0.0` = attend
- `torch.finfo(dtype).min` = block (avoids NaN issues vs -inf)

### All Tests Pass (28 total)

```
packages/data_handler/tests/test_attention_isolation.py - 5 passed
packages/data_handler/tests/test_sequence_packing.py - 23 passed
```

### Remaining Issue (Separate PR)

Gemini identified a domain tracking bug in `PackedDataset.__iter__` where domain labels get scrambled when `tokenize_batch_size > 1`. This is a pre-existing issue unrelated to the attention mask fix - will address in a separate PR.

**Comment by Lev-Stambler** [comment]
## Training Crash - Mask Shape Issue

Training crashed with:
```
attn_mask: size=(8, 1, 1024, 1024, 1024)  # 5D - WRONG!
```

Should be 4D `(batch, 1, seq, seq)`. Transformers is expanding our 2D mask incorrectly.

### Temporary Fix

Set `use_document_mask=False` as default in `PackedDataset.__init__()`. Training will work but without document isolation.

### Proper Fix (TODO)

Return mask as 4D tensor from dataset. The mask should be shaped `(1, 1, seq, seq)` per sample, so after batching it becomes `(batch, 1, seq, seq)`.

```python
# In _compute_attention_mask:
mask_2d = compute_document_attention_mask(input_ids, sep_id)  # (seq, seq)
return mask_2d.unsqueeze(0).unsqueeze(0)  # (1, 1, seq, seq)
```

This requires testing with the actual training pipeline to ensure transformers handles it correctly.

**Comment by Lev-Stambler** [comment]
## ✅ Fix Complete: 4D Mask Implementation

### The Fix
Return mask as 3D `(1, seq, seq)` per sample, so after DataLoader batching it becomes 4D `(batch, 1, seq, seq)`.

```python
mask_2d = compute_document_attention_mask(input_ids, sep_id)  # (seq, seq)
return mask_2d.unsqueeze(0)  # (1, seq, seq) → after batching: (batch, 1, seq, seq)
```

### Verified
1. **Shape test**: `attention_mask.shape = (2, 1, 16, 16)` ✓
2. **Model forward**: SmolLM-135M accepts the mask without errors ✓
3. **All tests pass**: 5/5 attention isolation tests ✓

### Changes
- `mixing.py`: Return 3D mask `(1, seq, seq)` instead of 2D
- `use_document_mask=True` re-enabled as default

Ready to merge! The sequence packing now properly isolates documents.
